src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1643,29 +1754,581 @@ protected Object parseValue(CommandSource source, CommandArgs args) throws Argum
+                    found = true;
                 }
             }
+            if (found) {
+                throw new IllegalArgumentException("Input value " + choice + " was not an entity of the required type!");
+            }
             throw new IllegalArgumentException("Input value " + choice + " was not an entity");
         }
 
-        private static Entity tryReturnSource(CommandSource source, CommandArgs args) throws ArgumentParseException {
-            if (source instanceof Entity) {
+        private Entity tryReturnSource(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            if (source instanceof Entity && this.checkEntity((Entity) source)) {
                 return (Entity) source;
-            } else if (source instanceof ProxySource && ((ProxySource) source).getOriginalSource() instanceof Entity) {
-                return (Entity) ((ProxySource) source).getOriginalSource();
+            }
+            if (source instanceof ProxySource) {
+                CommandSource proxy = ((ProxySource) source).getOriginalSource();
+                if (proxy instanceof Entity && this.checkEntity((Entity) proxy)) {
+                    return (Entity) proxy;
+                }
+            }
+            throw args.createError(t("No entities matched and source was not an entity!"));
+        }
+
+        private Entity tryReturnTarget(CommandSource source, CommandArgs args, Class<? extends Entity> type) throws ArgumentParseException {
+            Entity entity = tryReturnSource(source, args);
+            return entity.getWorld().getIntersectingEntities(entity, 10).stream()
+                    .filter(e -> !e.getEntity().equals(entity)).map(EntityUniverse.EntityHit::getEntity)
+                    .filter(this::checkEntity).findFirst()
+                    .orElseThrow(() -> args.createError(t("No entities matched and source was not looking at a valid entity!")));
+        }
+
+        private boolean checkEntity(Entity entity) {
+            return this.type == null || this.type.isAssignableFrom(entity.getClass());
+        }
+
+        @Override
+        public Text getUsage(CommandSource src) {
+            return src instanceof Entity && (this.returnSource || this.returnTarget) ? Text.of("[", this.getKey(), "]") : super.getUsage(src);
+        }
+    }
+
+    /**
+     * Expect an argument to represent a {@link URL}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement url(Text key) {
+        return new URLElement(key);
+    }
+
+    private static class URLElement extends KeyElement {
+
+        protected URLElement(Text key) {
+            super(key);
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            String str = args.next();
+            URL url;
+            try {
+                url = new URL(str);
```

Comment:
Nitpicking really (avoiding duplicate code!): but you could just write `return new URL(str).toURI();` and use a multicatch on the catch block given they give the same exception message.

Not stupidly important, just me being me here.